File: probRoot.py
#!/usr/local/bin/python2.7

#
# Use python wrapper to make 2D oscillagrams
# 

import numpy
from   math  import *

from BargerPropagator import *

from ROOT import TFile
from ROOT import TH2D
from ROOT import TCanvas
from ROOT import ROOT
import ROOT

#ROOT.gROOT.SetBatch(1)
ROOT.gStyle.SetOptStat(0)

# Create the method to compute probability
bNu = BargerPropagator()

# Binning	
NBinsEnergy    = 200
Dm12ZenithNbin = 200 

# Zenith Angle Range
Dm12ZenithEdge = numpy.zeros(Dm12ZenithNbin+1)
cz_start       = -1.001
cz_end         =  1.101
cz_step        = ( cz_end - cz_start)/float(Dm12ZenithNbin)

# Energy Range
EnergyBins     = numpy.zeros(NBinsEnergy+1)
e_start        = 0.110000001
e_end          = 300.0
e_step         = log10(e_end/e_start)/float(NBinsEnergy)
   
# Oscillation Parameters
kSquared = True  # are we using sin^2(x) variables?
kNuBar  =  1
DM2     = 2.5e-3
Theta23 = 0.5
Theta13 = 0.0238
dm2     = 7.9e-5
Theta12 = 0.302
#dcp     = 4.7123
dcp     = 0

# Print useful information
print "Using\tDM2\t", DM2, \
    "\n\tTheta23\t", Theta23, \
    "\n\tTheta13\t", Theta13, \
    "\n\tdm2\t", dm2, \
    "\n\tTheta12\t", Theta12, \
    "\n\tdcp\t", dcp 

print "From [", e_start, "-", e_end, "] GeV"

# Setup the axis for the histograms

for i in range(0, NBinsEnergy):
    Entry = e_start*pow( 10.0 , float(i)*e_step )
    EnergyBins[i] =Entry

EnergyBins[NBinsEnergy] = EnergyBins[NBinsEnergy-1]*1.001

Dm12ZenithEdge[0]= cz_start*0.9999
for i in range(1, Dm12ZenithNbin):
    Dm12ZenithEdge[i] = Dm12ZenithEdge[0] + float(i)*cz_step

Dm12ZenithEdge[Dm12ZenithNbin] = Dm12ZenithEdge[Dm12ZenithNbin-1]*1.001

# NuE Histograms
NuEToNuE3f     = TH2D("NuEToNuE3f","3 Flavor P_{#nu_{e}#rightarrow#nu_{e}}",
                      NBinsEnergy -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge)
NuEToNuMu3f    = TH2D("NuEToNuMu3f","3 Flavor P_{#nu_{e}#rightarrow#nu_{#mu}}",
                      NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge)
NuEToNuTau3f   = TH2D("NuEToNuTau3f","3 Flavor P_{#nu_{e}#rightarrow#nu_{#tau}}",
                      NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge)
NuEToNuX3f     = TH2D("NuEToNuX3f","3 Flavor P_{#nu_{e}#rightarrow#nu_{x}}",
                      NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge)
   
#NuMu Histograms
NuMuToNuE3f   = TH2D("NuMuToNuE3f","3 Flavor P_{#nu_{#mu}#rightarrow#nu_{e}}",
                     NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge)
NuMuToNuMu3f  = TH2D("NuMuToNuMu3f","3 Flavor P_{#nu_{#mu}#rightarrow#nu_{#mu}}",
                     NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge)
NuMuToNuTau3f = TH2D("NuMuToNuTau3f","3 Flavor P_{#nu_{#mu}#rightarrow#nu_{#tau}}",
                     NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge)
NuMuToNuX3f   = TH2D("NuMuToNuX3f","3 Flavor P_{#nu_{#mu}#rightarrow#nu_{x}}",
                     NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge)

#NuTau Histograms   
NuTauToNuE3f  = TH2D("NuTauToNuE3f","3 Flavor P_{#nu_{#tau}#rightarrow#nu_{e}}",
                     NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge)
NuTauToNuMu3f = TH2D("NuTauToNuMu3f","3 Flavor P_{#nu_{#tau}#rightarrow#nu_{#mu}}",
                     NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge)
NuTauToNuTau3f= TH2D("NuTauToNuTau3f","3 Flavor P_{#nu_{#tau}#rightarrow#nu_{#tau}}",
                     NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge)
NuTauToNuX3f  = TH2D("NuTauToNuX3f","3 Flavor P_{#nu_{#tau}#rightarrow#nu_{x}}",
                     NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge)

# Two Flavor Histogram                                
NuMuToNuTau2f = TH2D("NuMuToNuTau2f","2 Flavor P_{#nu_{#mu}#rightarrow#nu_{#tau}}",
                     NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge)

# Fill the Histograms
for i in range(0, NBinsEnergy):
    energy = e_start*pow(10.0, float(i)*e_step)
    for j in range(0, Dm12ZenithNbin):
        
        cosineZ = cz_start + float(j)*cz_step
  
        bNu.SetMNS( Theta12,  Theta13, Theta23, dm2, DM2, dcp , energy, kSquared, kNuBar ) 
        bNu.DefinePath( cosineZ, 25.00, True  )
        bNu.propagate( 1*kNuBar )
  
        NuEToNuE3f.Fill( energy, cosineZ, bNu.GetProb(1,1) ) 
        NuEToNuMu3f.Fill( energy, cosineZ, bNu.GetProb(1,2) ) 
        NuEToNuTau3f.Fill( energy, cosineZ, bNu.GetProb(1,3) ) 
        NuEToNuX3f.Fill( energy, cosineZ, 1.0 - bNu.GetProb(1,1) ) 

        NuMuToNuE3f.Fill( energy, cosineZ, bNu.GetProb(2,1) ) 
        NuMuToNuMu3f.Fill( energy, cosineZ, bNu.GetProb(2,2) ) 
        NuMuToNuTau3f.Fill( energy, cosineZ, bNu.GetProb(2,3) ) 
        NuMuToNuX3f.Fill( energy, cosineZ, 1.0 - bNu.GetProb(2,2) ) 

        NuTauToNuE3f.Fill( energy, cosineZ, bNu.GetProb(3,1) ) 
        NuTauToNuMu3f.Fill( energy, cosineZ, bNu.GetProb(3,2) ) 
        NuTauToNuTau3f.Fill( energy, cosineZ, bNu.GetProb(3,3) ) 
        NuTauToNuX3f.Fill( energy, cosineZ, 1.0 - bNu.GetProb(3,3) ) 

tmp = TFile("RawProb.root", "recreate")

NuEToNuE3f.Write()
NuEToNuMu3f.Write()
NuEToNuTau3f.Write()
NuEToNuX3f.Write()
   
NuMuToNuE3f.Write()
NuMuToNuMu3f.Write()
NuMuToNuTau3f.Write()
NuMuToNuX3f.Write()

NuTauToNuE3f.Write()
NuTauToNuMu3f.Write()
NuTauToNuTau3f.Write()
NuTauToNuX3f.Write()

NuMuToNuTau2f.Write()

tmp.Close()
   	
print "Happy oscillating! Job is finished."

File: probRoot.cc
#include <iostream>

#include "BargerPropagator.h"

#include "TFile.h"
#include "TH2D.h"

int main(int argc, char * argv[] )
{

   double dcp_in =  0.;
   double h_in   = 1.0;
   int    v_in   = 1.0;

   if( argc >= 2 ) dcp_in = (double) atof( argv[1] );
   if( argc >= 3 ) h_in   = (double) atof( argv[2] );
   if( argc >= 4 ) v_in   = (int)    atoi( argv[3] );

   h_in = ( h_in > 0 ? 1.0 : -1.0 );

   double cosineZ, energy;
   double e_start, e_end, e_step, cz_start, cz_end, cz_step;
   int i, j ;

//// Types of Averaging	
   int NBinsEnergy = 200;
   int Dm12ZenithNbin = 200; 


// Zenith Angle Range
   double Dm12ZenithEdge[Dm12ZenithNbin+1];
   cz_start = -1.001;
   cz_end = 1.101;
   cz_step = ( cz_end - cz_start)/double(Dm12ZenithNbin);


// Energy Range
   double EnergyBins[NBinsEnergy+1];
   e_start = 0.110000001;
   e_end = 300.;
   e_step = log10(e_end/e_start)/double(NBinsEnergy);

   
/// Oscillation Parameters
// bool kSquared = false;   // are we using sin^2(x) variables?
   bool kSquared = true ;   // are we using sin^2(x) variables?

   int    kNuBar  =  1 * v_in;
   double DM2     =  h_in * 2.5e-3;
   double Theta23 =  0.5    ;
   double Theta13 =  0.0215 ;
   double dm2     =  7.6e-5 ;
   double Theta12 =  0.302  ;
// double delta   = 270.0 * (3.1415926/180.0);
   double delta   =  dcp_in * (3.1415926/180.0);


   std::cout << "Using          " << std::endl
             << "      DM2      " <<  DM2      << std::endl
             << "      Theta23  " <<  Theta23  << std::endl
             << "      Theta13  " <<  Theta13  << std::endl
             << "      dm2      " <<  dm2      << std::endl
             << "      Theta12  " <<  Theta12  << std::endl
             << "      delta    " <<  delta    << std::endl;
   std::cout << "      knubar   " <<  kNuBar   << std::endl;

   std::cout << "From "
	     << " [ " << e_start << " - " << e_end << " ] GeV " << endl;
   
// Methods to Compute Probability
   NeutrinoPropagator * myNu;
   BargerPropagator   * bNu; 

   bNu = new BargerPropagator( true );
   //bNu = new BargerPropagator("full.PREM.dat");

   bNu->UseMassEigenstates( false ); // this is default (use flavor eigenstates)

   // Octant for Theta23 in sin2(2x) mode
   bNu->SetDefaultOctant( 23, 2 );
   // use the standard barger
   myNu = bNu;

   double Entry = e_start;
   for(i=0; i<NBinsEnergy; i++ )
        {
                Entry = e_start*pow( 10.0 , double(i)*e_step );
                EnergyBins[i] = Entry;
   }
   EnergyBins[NBinsEnergy] = EnergyBins[NBinsEnergy-1]*1.001;

   
   Dm12ZenithEdge[0]= cz_start*0.9999;
   for ( i=1; i<Dm12ZenithNbin ; i++ )
   	Dm12ZenithEdge[i] = Dm12ZenithEdge[0] + double(i)*cz_step;

   Dm12ZenithEdge[Dm12ZenithNbin] = Dm12ZenithEdge[Dm12ZenithNbin-1]*1.001;

   
   TH2D *NuEToNuE3f 	=  new TH2D("NuEToNuE3f","3 Flavor P_{#nu_{e}#rightarrow#nu_{e}}",
   					NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge);
   TH2D *NuEToNuMu3f 	=  new TH2D("NuEToNuMu3f","3 Flavor P_{#nu_{e}#rightarrow#nu_{#mu}}",
   					NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge);
   TH2D *NuEToNuTau3f 	=  new TH2D("NuEToNuTau3f","3 Flavor P_{#nu_{e}#rightarrow#nu_{#tau}}",
   					NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge);
   TH2D *NuEToNuX3f 	=  new TH2D("NuEToNuX3f","3 Flavor P_{#nu_{e}#rightarrow#nu_{x}}",
   					NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge);
   

   
   TH2D *NuMuToNuE3f 	=  new TH2D("NuMuToNuE3f","3 Flavor P_{#nu_{#mu}#rightarrow#nu_{e}}",
   					NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge);
   TH2D *NuMuToNuMu3f 	=  new TH2D("NuMuToNuMu3f","3 Flavor P_{#nu_{#mu}#rightarrow#nu_{#mu}}",
   					NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge);
   TH2D *NuMuToNuTau3f 	=  new TH2D("NuMuToNuTau3f","3 Flavor P_{#nu_{#mu}#rightarrow#nu_{#tau}}",
   					NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge);
   TH2D *NuMuToNuX3f 	=  new TH2D("NuMuToNuX3f","3 Flavor P_{#nu_{#mu}#rightarrow#nu_{x}}",
   					NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge);

   
   TH2D *NuTauToNuE3f 	=  new TH2D("NuTauToNuE3f","3 Flavor P_{#nu_{#tau}#rightarrow#nu_{e}}",
   					NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge);
   TH2D *NuTauToNuMu3f 	=  new TH2D("NuTauToNuMu3f","3 Flavor P_{#nu_{#tau}#rightarrow#nu_{#mu}}",
   					NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge);
   TH2D *NuTauToNuTau3f 	=  new TH2D("NuTauToNuTau3f","3 Flavor P_{#nu_{#tau}#rightarrow#nu_{#tau}}",
   					NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge);
   TH2D *NuTauToNuX3f 	=  new TH2D("NuTauToNuX3f","3 Flavor P_{#nu_{#tau}#rightarrow#nu_{x}}",
   					NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge);


   
   TH2D *NuMuToNuTau2f 	=  new TH2D("NuMuToNuTau2f","2 Flavor P_{#nu_{#mu}#rightarrow#nu_{#tau}}",
   					NBinsEnergy  -1 , EnergyBins, Dm12ZenithNbin -1, Dm12ZenithEdge);

//------------------- End of Raw Probability Plots	


   double total = 0.0;
   // fill the probabilities
   for ( i = 0 ; i <= NBinsEnergy ; i ++ ) 
   {

     energy = e_start*pow(10.0, double(i)*e_step);
     for ( j = 0 ; j <= Dm12ZenithNbin ; j++ )
     { 

        cosineZ = cz_start + double(j)*cz_step;
  
        myNu->SetMNS( Theta12,  Theta13, Theta23, dm2, DM2, delta , energy, kSquared, kNuBar ); 

//      if( i == 0 && j == 0 )
//        print_mixing_matrix();

        myNu->DefinePath( cosineZ, 25.00  );
        myNu->propagate( 1*kNuBar );
  
        total = 0.0;
	total += myNu->GetProb(1,1);
	total += myNu->GetProb(1,2);
	total += myNu->GetProb(1,3);

	if( fabs( 1.00 - total ) > 1.0e-7 ) abort();
	
	
        NuEToNuE3f 	->Fill( energy, cosineZ, myNu->GetProb(1,1) ) ;
        NuEToNuMu3f 	->Fill( energy, cosineZ, myNu->GetProb(1,2) ) ;
        NuEToNuTau3f 	->Fill( energy, cosineZ, myNu->GetProb(1,3) ) ;
        NuEToNuX3f 	->Fill( energy, cosineZ, 1.0 - myNu->GetProb(1,1) ) ;

        NuMuToNuE3f 	->Fill( energy, cosineZ, myNu->GetProb(2,1) ) ;
        NuMuToNuMu3f 	->Fill( energy, cosineZ, myNu->GetProb(2,2) ) ;
        NuMuToNuTau3f 	->Fill( energy, cosineZ, myNu->GetProb(2,3) ) ;
        NuMuToNuX3f 	->Fill( energy, cosineZ, 1.0 - myNu->GetProb(2,2) ) ;

        NuTauToNuE3f 	->Fill( energy, cosineZ, myNu->GetProb(3,1) ) ;
        NuTauToNuMu3f 	->Fill( energy, cosineZ, myNu->GetProb(3,2) ) ;
        NuTauToNuTau3f 	->Fill( energy, cosineZ, myNu->GetProb(3,3) ) ;
        NuTauToNuX3f 	->Fill( energy, cosineZ, 1.0 - myNu->GetProb(3,3) ) ;
     }// End of Cosine Z Looping //
   
   } // End Energy Loop //


   TFile *tmp = new TFile("RawProb.root", "recreate");
   tmp->cd();

   NuEToNuE3f    ->Write();
   NuEToNuMu3f   ->Write();
   NuEToNuTau3f  ->Write();
   NuEToNuX3f    ->Write();
   
   NuMuToNuE3f   ->Write();
   NuMuToNuMu3f  ->Write();
   NuMuToNuTau3f ->Write();
   NuMuToNuX3f   ->Write();

   NuTauToNuE3f  ->Write();
   NuTauToNuMu3f ->Write();
   NuTauToNuTau3f->Write();
   NuTauToNuX3f  ->Write();

   NuMuToNuTau2f ->Write();

   tmp->Close();
   	
   std::cout << std::endl<<"Done Cowboy!" << std::endl;
   return 0;
}

